src/librustc/traits/select.rs

```diff
@@ -978,7 +1002,9 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
         // not update) the cache.
         let recursion_limit = self.infcx.tcx.sess.recursion_limit.get();
         if stack.obligation.recursion_depth >= recursion_limit {
-            self.infcx().report_overflow_error(&stack.obligation, true);
+            // FIXME: not reporting the error eagerly means it'll eventually get reported for
+            // the root obligation, not the current stack.obligation
+            return Err(Overflow);
```

Comment:
This results in UI errors?

Reply:
I believe so (of the sort in `test/ui/did_you_mean/recursion_limit.rs`, see the [latest Travis run](https://travis-ci.org/rust-lang/rust/jobs/361148796#L1689)), but now I'm less sure because of the [bug you pointed out](https://github.com/rust-lang/rust/pull/48995#discussion_r179209238). Will double-check. But don't you think that it would result in UI errors? AFAICT we are changing the precise obligation for which we are reporting the error.